File: src/main/java/jp/kentan/minecraft/neko_event/config/provider/DungeonConfigProvider.java
package jp.kentan.minecraft.neko_event.config.provider;

import jp.kentan.minecraft.neko_event.config.ConfigManager;
import jp.kentan.minecraft.neko_event.config.listener.ConfigListener;
import jp.kentan.minecraft.neko_event.game.dungeon.model.Dungeon;
import jp.kentan.minecraft.neko_event.util.Log;
import org.apache.commons.lang.time.DateUtils;
import org.bukkit.Bukkit;
import org.bukkit.Location;
import org.bukkit.configuration.file.FileConfiguration;
import org.bukkit.configuration.file.YamlConfiguration;

import java.io.File;
import java.util.*;

import static jp.kentan.minecraft.neko_event.util.NekoUtil.DATE_FORMAT;

public class DungeonConfigProvider {

    private static File sConfigFile;

    private static ConfigListener<Dungeon> sListener;

    public static void setup(File dataFolder){
        sConfigFile = new File(dataFolder + File.separator + "dungeon.yml");

        ConfigManager.createIfNeed(sConfigFile);
    }

    public static void bindListener(ConfigListener<Dungeon> listener){
        sListener = listener;
    }

    public static void load(){
        try {
            FileConfiguration config = new YamlConfiguration();
            config.load(sConfigFile);

            final Map<String, Dungeon> dungeonMap = new HashMap<>();

            if(!config.isConfigurationSection("Dungeon")){
                Log.warn("Dungeon section not found.");
                return;
            }

            final Set<String> dungeonIdSet = config.getConfigurationSection("Dungeon").getKeys(false);

            dungeonIdSet.forEach(id -> {
                final String path = "Dungeon." + id;

                Dungeon dungeon = new Dungeon(id,
                        config.getString(path + ".name", "ダンジョン"),
                        config.isConfigurationSection(path + ".Join.Location") ?
                                new Location(
                                        Bukkit.getWorld(config.getString(path + ".Join.Location.world")),
                                        config.getDouble(path + ".Join.Location.x"),
                                        config.getDouble(path + ".Join.Location.y"),
                                        config.getDouble(path + ".Join.Location.z"),
                                        (float)config.getDouble(path + ".Join.Location.yaw"),
                                        (float)config.getDouble(path + ".Join.Location.pitch")
                                ) : null,
                        config.isConfigurationSection(path + ".Clear.Location") ?
                                new Location(
                                        Bukkit.getWorld(config.getString(path + ".Clear.Location.world")),
                                        config.getDouble(path + ".Clear.Location.x"),
                                        config.getDouble(path + ".Clear.Location.y"),
                                        config.getDouble(path + ".Clear.Location.z"),
                                        (float)config.getDouble(path + ".Clear.Location.yaw"),
                                        (float)config.getDouble(path + ".Clear.Location.pitch")
                                ) : null,
                        config.getString(path + ".Join.playerMsg"),
                        config.getString(path + ".Join.broadcastMsg"),
                        config.getString(path + ".Clear.playerMsg"),
                        config.getString(path + ".Clear.broadcastMsg"),
                        config.getString(path + ".Clear.titleText"),
                        config.getBoolean(path + ".Clear.enableSound"),
                        config.getString(path + ".TimedOut.broadcastMsg"),
                        config.getInt(path + ".Reward.ticketAmount"),
                        config.getString(path + ".Reward.gachaId")
                );

                dungeonMap.put(id, dungeon);
            });

            Log.info(dungeonMap.size() + "個のダンジョンを読み込みました.");

            sListener.onUpdate(dungeonMap);
        } catch (Exception e) {
            e.printStackTrace();
            Log.error(e.getMessage());
        }
    }

    public static Dungeon create(String id, String name, Location joinLocation, Location clearLocation, String joinPlayerMsg,
                                           String joinBroadcastMsg, String clearPlayerMsg, String clearBroadcastMsg, String clearTitleText,
                                           Boolean enableClearSound, String timedOutBroadcastMsg, int rewardTicketAmount, String rewordGachaId) {
        final String path = "Dungeon." + id;

        try {
            FileConfiguration config = new YamlConfiguration();
            config.load(sConfigFile);

            if(config.isConfigurationSection(path)){
                Log.error("ダンジョン({id})は既に登録されています.".replace("{id}",id));
                return null;
            }
        } catch (Exception e) {
            e.printStackTrace();
            Log.error(e.getMessage());
            return null;
        }

        if(!save(new LinkedHashMap<String, Object>(){
            {
                put(path + ".name", name);

                if(joinLocation != null) {
                    put(path + ".Join.Location.world", joinLocation.getWorld().getName());
                    put(path + ".Join.Location.x", joinLocation.getX());
                    put(path + ".Join.Location.y", joinLocation.getY());
                    put(path + ".Join.Location.z", joinLocation.getZ());
                    put(path + ".Join.Location.yaw", (double) joinLocation.getYaw());
                    put(path + ".Join.Location.pitch", (double) joinLocation.getPitch());
                }

                if(clearLocation != null) {
                    put(path + ".Clear.Location.world", clearLocation.getWorld().getName());
                    put(path + ".Clear.Location.x", clearLocation.getX());
                    put(path + ".Clear.Location.y", clearLocation.getY());
                    put(path + ".Clear.Location.z", clearLocation.getZ());
                    put(path + ".Clear.Location.yaw", (double) clearLocation.getYaw());
                    put(path + ".Clear.Location.pitch", (double) clearLocation.getPitch());
                }

                put(path + ".Join.playerMsg", joinPlayerMsg);
                put(path + ".Join.broadcastMsg", joinBroadcastMsg);
                put(path + ".Clear.playerMsg", clearPlayerMsg);
                put(path + ".Clear.broadcastMsg", clearBroadcastMsg);
                put(path + ".Clear.titleText", clearTitleText);
                put(path + ".Clear.enableSound", enableClearSound);
                put(path + ".TimedOut.broadcastMsg", timedOutBroadcastMsg);

                put(path + ".Reward.ticketAmount", rewardTicketAmount);
                put(path + ".Reward.gachaId", rewordGachaId);
            }
        })){
            return null;
        }

        return new Dungeon(id, name, joinLocation, clearLocation, joinPlayerMsg, joinBroadcastMsg, clearPlayerMsg, clearBroadcastMsg, clearTitleText, enableClearSound, timedOutBroadcastMsg, rewardTicketAmount, rewordGachaId);
    }

    public static boolean update(String id, Location joinLocation, Location clearLocation, String joinPlayerMsg, String joinBroadcastMsg, String clearPlayerMsg,
                                        String clearBroadcastMsg, String clearTitleText, Boolean enableClearSound, String timedOutBroadcastMsg, int rewardTicketAmount, String rewardGachaId) {
        final String path = "Dungeon." + id;

        try {
            FileConfiguration config = new YamlConfiguration();
            config.load(sConfigFile);

            if(!config.isConfigurationSection(path)){
                Log.error("ダンジョン({id})は存在しません.".replace("{id}",id));
                return false;
            }
        } catch (Exception e) {
            e.printStackTrace();
            Log.error(e.getMessage());
            return false;
        }

        return save(new LinkedHashMap<String, Object>(){
            {
                if(joinLocation != null) {
                    put(path + ".Join.Location.world", joinLocation.getWorld().getName());
                    put(path + ".Join.Location.x", joinLocation.getX());
                    put(path + ".Join.Location.y", joinLocation.getY());
                    put(path + ".Join.Location.z", joinLocation.getZ());
                    put(path + ".Join.Location.yaw", (double) joinLocation.getYaw());
                    put(path + ".Join.Location.pitch", (double) joinLocation.getPitch());
                }

                if(clearLocation != null){
                    put(path + ".Clear.Location.world", clearLocation.getWorld().getName());
                    put(path + ".Clear.Location.x", clearLocation.getX());
                    put(path + ".Clear.Location.y", clearLocation.getY());
                    put(path + ".Clear.Location.z", clearLocation.getZ());
                    put(path + ".Clear.Location.yaw", (double)clearLocation.getYaw());
                    put(path + ".Clear.Location.pitch", (double)clearLocation.getPitch());
                }

                if(joinPlayerMsg != null) {
                    put(path + ".Join.playerMsg", joinPlayerMsg);
                }
                if(joinBroadcastMsg != null) {
                    put(path + ".Join.broadcastMsg", joinBroadcastMsg);
                }

                if(clearPlayerMsg != null) {
                    put(path + ".Clear.playerMsg", clearPlayerMsg);
                }
                if(clearBroadcastMsg != null) {
                    put(path + ".Clear.broadcastMsg", clearBroadcastMsg);
                }
                if(clearTitleText != null){
                    put(path + ".Clear.titleText", clearTitleText);
                }
                if(enableClearSound != null){
                    put(path + ".Clear.enableSound", enableClearSound);
                }

                if(timedOutBroadcastMsg != null){
                    put(path + ".TimedOut.broadcastMsg", timedOutBroadcastMsg);
                }

                if(rewardTicketAmount > 0) {
                    put(path + ".Reward.ticketAmount", rewardTicketAmount);
                }
                if(rewardGachaId != null) {
                    put(path + ".Reward.gachaId", rewardGachaId);
                }
            }
        });
    }

    public static boolean delete(String id) {
        final String path = "Dungeon." + id;

        return save(new HashMap<String, Object>(){
            {
                put(path , null);
            }
        });
    }

    public static boolean saveClearDate(UUID uuid, String id){
        return PlayerConfigProvider.save(uuid, new HashMap<String, Object>(){
            {
                put("Dungeon." + id + ".clearDate", DATE_FORMAT.format(new Date()));
            }
        });
    }

    public static boolean hasClearedToday(UUID uuid, String id){
        try {
            final String strPrev = (String) PlayerConfigProvider.get(uuid, "Dungeon." + id + ".clearDate", null);

            if(strPrev == null){
                return false;
            }

            final Date prev = DATE_FORMAT.parse(strPrev);
            final Date today = DateUtils.truncate(new Date(), Calendar.DAY_OF_MONTH);

            return prev != null && prev.compareTo(today) == 0;
        }catch (Exception e){
            e.printStackTrace();
            Log.error(e.getMessage());
            return false;
        }
    }

    private static boolean save(Map<String, Object> dataList) {
        try {
            FileConfiguration config = new YamlConfiguration();
            config.load(sConfigFile);

            dataList.forEach(config::set);

            config.save(sConfigFile);
        } catch (Exception e) {
            e.printStackTrace();
            Log.error(e.getMessage());

            return false;
        }
        return true;
    }
}

File: src/main/java/jp/kentan/minecraft/nekoevent/config/provider/DungeonConfigProvider.kt
package jp.kentan.minecraft.nekoevent.config.provider

import jp.kentan.minecraft.nekoevent.component.model.Dungeon
import jp.kentan.minecraft.nekoevent.config.ConfigManager.Companion.DATE_FORMAT
import jp.kentan.minecraft.nekoevent.config.ConfigUpdateListener
import jp.kentan.minecraft.nekoevent.util.Log
import org.apache.commons.lang.time.DateUtils
import org.bukkit.Bukkit
import org.bukkit.Location
import org.bukkit.configuration.file.YamlConfiguration
import org.bukkit.entity.Player
import java.io.File
import java.util.*

class DungeonConfigProvider(
        dataFolder: File,
        private val playerConfig: PlayerConfigProvider
) : BaseConfigProvider(dataFolder, "dungeon.yml") {

    var listener: ConfigUpdateListener<Dungeon>? = null

    fun load() {
        try {
            val config = YamlConfiguration.loadConfiguration(configFile)

            if (!config.isConfigurationSection("Dungeon")) {
                Log.warn("Dungeonセクションがありません.")
                return
            }

            val dungeonIdSet = config.getConfigurationSection("Dungeon").getKeys(false)
            val dungeonMap = dungeonIdSet.associate { id ->
                val path = "Dungeon.$id"

                val dungeon = Dungeon(
                        id,
                        config.getString("$path.name", "ダンジョン"),
                        config.getInt("$path.Reward.ticketAmount"),
                        config.getString("$path.Reward.gachaId"),
                        if (config.isConfigurationSection("$path.Join.Location"))
                            Location(
                                    Bukkit.getWorld(config.getString("$path.Join.Location.world")),
                                    config.getDouble("$path.Join.Location.x"),
                                    config.getDouble("$path.Join.Location.y"),
                                    config.getDouble("$path.Join.Location.z"),
                                    config.getDouble("$path.Join.Location.pitch").toFloat(),
                                    config.getDouble("$path.Join.Location.yaw").toFloat()
                            )
                        else
                            null,
                        config.getString("$path.Join.message"),
                        config.getString("$path.Join.broadcastMessage"),
                        if (config.isConfigurationSection("$path.Clear.Location"))
                            Location(
                                    Bukkit.getWorld(config.getString("$path.Clear.Location.world")),
                                    config.getDouble("$path.Clear.Location.x"),
                                    config.getDouble("$path.Clear.Location.y"),
                                    config.getDouble("$path.Clear.Location.z"),
                                    config.getDouble("$path.Clear.Location.pitch").toFloat(),
                                    config.getDouble("$path.Clear.Location.yaw").toFloat()
                            )
                        else
                            null,
                        config.getString("$path.Clear.message"),
                        config.getString("$path.Clear.broadcastMessage"),
                        config.getBoolean("$path.Clear.enabledSound"),
                        config.getString("$path.Lock.message"),
                        config.getString("$path.Lock.broadcastMessage"),
                        config.getString("$path.Unlock.broadcastMessage")
                )

                return@associate id to dungeon
            }

            listener?.onConfigUpdate(dungeonMap)
        } catch (e: Exception) {
            e.printStackTrace()
            Log.error(e)
        }
    }

    fun update(dungeon: Dungeon): Boolean {
        val path = "Dungeon.${dungeon.id}"
        val dataMap = LinkedHashMap<String, Any?>().apply {
            put("$path.name", dungeon.name)

            put("$path.Reward.ticketAmount", dungeon.rewardTicketAmount)
            put("$path.Reward.gachaId", dungeon.rewardGachaId)

            if (dungeon.joinLocation != null) {
                val loc = dungeon.joinLocation
                put("$path.Join.Location.world", loc.world.name)
                put("$path.Join.Location.x", loc.x)
                put("$path.Join.Location.y", loc.y)
                put("$path.Join.Location.z", loc.z)
                put("$path.Join.Location.yaw", loc.yaw)
                put("$path.Join.Location.pitch", loc.pitch)
            } else {
                put("$path.Join.Location", null)
            }
            put("$path.Join.message", dungeon.joinMessage)
            put("$path.Join.broadcastMessage", dungeon.joinBroadcastMessage)

            if (dungeon.clearLocation != null) {
                val loc = dungeon.clearLocation
                put("$path.Clear.Location.world", loc.world.name)
                put("$path.Clear.Location.x", loc.x)
                put("$path.Clear.Location.y", loc.y)
                put("$path.Clear.Location.z", loc.z)
                put("$path.Clear.Location.yaw", loc.yaw)
                put("$path.Clear.Location.pitch", loc.pitch)
            } else {
                put("$path.Clear.Location", null)
            }
            put("$path.Clear.message", dungeon.clearMessage)
            put("$path.Clear.broadcastMessage", dungeon.clearBroadcastMessage)
            put("$path.Clear.enabledSound", dungeon.enabledClearSound)

            put("$path.Lock.message", dungeon.lockMessage)
            put("$path.Lock.broadcastMessage", dungeon.lockBroadcastMessage)

            put("$path.Unlock.broadcastMessage", dungeon.unlockBroadcastMessage)
        }

        return super.save(dataMap)
    }

    fun delete(keyId: String): Boolean {
        return super.save(mapOf("Dungeon.$keyId" to null))
    }

    fun hasClearedToday(player: Player, dungeon: Dungeon): Boolean {
        try {
            val oldDate = playerConfig.get(player.uniqueId, "Dungeon.${dungeon.id}.clearDate") as String? ?: return false

            val prev = DATE_FORMAT.parse(oldDate)
            val today = DateUtils.truncate(Date(), Calendar.DAY_OF_MONTH)

            return prev.compareTo(today) == 0
        } catch (e: Exception) {
            e.printStackTrace()
            Log.error(e)
        }

        return false
    }

    fun saveClearDate(player: Player, dungeon: Dungeon) {
        playerConfig.save(player.uniqueId, mapOf("Dungeon.${dungeon.id}.clearDate" to DATE_FORMAT.format(Date())))
    }
}